Refactor Scrollbar to hold NSScrollerImp

nmoucht opened the pull request:
#### fdbd1e818def7f92e7c7e410229619c18d419d08
<pre>
Refactor Scrollbar to hold NSScrollerImp
<a href="https://bugs.webkit.org/show_bug.cgi?id=264014">https://bugs.webkit.org/show_bug.cgi?id=264014</a>
<a href="https://rdar.apple.com/117770817">rdar://117770817</a>

Reviewed by Simon Fraser.

Decouple creation of NSScrollerImps from the global map in
ScrollbarThemeMac. Instead, create ScrollbarMac to hold a reference to the
NSScrollerImp, and change the map to a set of scrollbars (which is still
necessary due to the WebScrollbarPrefsObserver stuff).

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::createNativeScrollbar):
* Source/WebCore/platform/mac/ScrollbarMac.h: Added.
(WebCore::ScrollbarMac::~ScrollbarMac):
* Source/WebCore/platform/mac/ScrollbarMac.mm: Added.
(WebCore::ScrollbarMac::ScrollbarMac):
(WebCore::ScrollbarMac::painterForScrollbar):
(WebCore::ScrollbarMac::createNSScrollerImp):
(WebCore::macScrollbarTheme):
(WebCore::ScrollbarMac::updateNSScrollerImpState):
(WebCore::ScrollbarMac::setNewPainterForScrollbar):
* Source/WebCore/platform/mac/ScrollbarThemeMac.h:
* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::scrollbarMap):
(+[WebScrollbarPrefsObserver appearancePrefsChanged:]):
(WebCore::ScrollbarThemeMac::registerScrollbar):
(WebCore::ScrollbarThemeMac::unregisterScrollbar):
(WebCore::ScrollbarThemeMac::painterForScrollbar):
(WebCore::ScrollbarThemeMac::hasThumb):
(WebCore::ScrollbarThemeMac::minimumThumbLength):
(WebCore::ScrollbarThemeMac::updateEnabledState):
(WebCore::paintScrollbar):
(WebCore::ScrollbarThemeMac::setNewPainterForScrollbar): Deleted.
* Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
(WebCore::ScrollbarsControllerMac::updateScrollerStyle):

Canonical link: <a href="https://commits.webkit.org/272044@main">https://commits.webkit.org/272044@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eeaed4a80a769a95012315bc76f88ed809dde5b7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30327 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9000 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31948 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32836 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27446 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/31026 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11198 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6233 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27415 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30635 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7561 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27205 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6480 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6627 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27052 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34176 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27648 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27554 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32816 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6606 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4761 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30629 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8361 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7353 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3939 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7147 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/mac/ScrollbarMac.h:
```diff
@@ -0,0 +1,52 @@
+
+    void createNSScrollerImp();
+    NSScrollerImp* painterForScrollbar();
+    void setNewPainterForScrollbar(RetainPtr<NSScrollerImp>&&);
```
The "forScrollbar" in these names is now redundant. And do we just treat the imps as "painters" or do we use them for other purposes?

I think just `scrollerImp()`/`setScrollerImp()` would be fine.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/mac/ScrollbarMac.h:
```diff
@@ -0,0 +1,52 @@
+
+private:
+    RetainPtr<NSScrollerImp> m_scrollerImp;
+
```
Extra blank line.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/mac/ScrollbarMac.h:
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (C) 2008 Apple Inc. All Rights Reserved.
```
2023

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/mac/ScrollbarMac.h:
```diff
@@ -0,0 +1,52 @@
+#if PLATFORM(MAC)
+
+#include "Scrollbar.h"
+#include "ScrollbarThemeComposite.h"
```
Not needed?

nmoucht replied:
Removed.

smfr commented on Source/WebCore/platform/mac/ScrollbarMac.h:
```diff
@@ -0,0 +1,52 @@
+    void createNSScrollerImp();
+    NSScrollerImp* painterForScrollbar();
+    void setNewPainterForScrollbar(RetainPtr<NSScrollerImp>&&);
+    void updateNSScrollerImpState();
```
Should be private.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/mac/ScrollbarMac.h:
```diff
@@ -0,0 +1,52 @@
+    ScrollbarMac(ScrollableArea&, ScrollbarOrientation, ScrollbarWidth);
+    ~ScrollbarMac() { }
+
+    void createNSScrollerImp();
```
Should be private.

nmoucht replied:
This is now being used by ScrollbarsControllerMac so keeping public.

smfr commented on Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
```diff
@@ -53,11 +54,11 @@
 namespace WebCore {
 
-using ScrollbarToScrollerImpMap = HashMap<Scrollbar*, RetainPtr<NSScrollerImp>>;
+using ScrollbarSet = HashSet<Scrollbar*>;
```
Our memory safe coding rules would say that this should be a hash of weak or checked pointers.

nmoucht replied:
Changed to checked pointer.

smfr commented on Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
```diff
@@ -957,8 +958,7 @@ - (void)invalidate
         auto newVerticalPainter = retainPtr([NSScrollerImp scrollerImpWithStyle:newStyle controlSize:(NSControlSize)verticalScrollbar->widthStyle() horizontal:NO replacingScrollerImp:oldVerticalPainter]);
```
I think you should strive to only have one call site that creates `NSScrollerImp` objects. Having one here, and one in ScrollbarMac forces me to think about how the arguments differ.

nmoucht replied:
Right, changed this to use createScrollerImp and pass the old NSScrollerImp.

smfr commented on Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
```diff
@@ -957,8 +958,7 @@ - (void)invalidate
         [m_scrollerImpPair setVerticalScrollerImp:newVerticalPainter.get()];
-        macTheme->setNewPainterForScrollbar(*verticalScrollbar, WTFMove(newVerticalPainter));
-        macTheme->didCreateScrollerImp(*verticalScrollbar);
+        static_cast<ScrollbarMac*>(verticalScrollbar)->setNewPainterForScrollbar(WTFMove(newVerticalPainter));
```
Would feel a bit nicer if we could `dynamicDowncast<>`

nmoucht replied:
Added isMacScrollbar so this is possible.

smfr commented on Source/WebCore/platform/Scrollbar.h:
```diff
@@ -196,3 +198,7 @@ class Scrollbar : public Widget {
 
 SPECIALIZE_TYPE_TRAITS_WIDGET(Scrollbar, isScrollbar())
 
+#define SPECIALIZE_TYPE_TRAITS_MAC_SCROLLBAR(ToValueTypeName, predicate) \
```
The `MAC` should not exist in this macro name.

nmoucht replied:
Changed to SPECIALIZE_TYPE_TRAITS_SCROLLBAR_HOLDS_SCROLLER_IMP.

smfr commented on Source/WebCore/platform/mac/ScrollbarMac.h:
```diff
@@ -0,0 +1,55 @@
+    ScrollbarMac(ScrollableArea&, ScrollbarOrientation, ScrollbarWidth);
+    ~ScrollbarMac() { }
+
+    NSScrollerImp* scrollerImp();
```
`scrollerImp() const`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/mac/ScrollbarMac.mm:
```diff
@@ -0,0 +1,74 @@
+void ScrollbarMac::updateScrollerImpState()
+{
+    if (auto* scrollbarTheme = macScrollbarTheme())
+        scrollbarTheme->didCreateScrollerImp(*this);
```
Need to fix ScrollbarTheme so this looks like `scrollbarTheme->didCreateScrollbar(*this)`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
```diff
@@ -155,29 +155,19 @@ + (void)registerAsObserver
+    scrollbarMap().take(&scrollbar);
 }
 
 NSScrollerImp *ScrollbarThemeMac::painterForScrollbar(Scrollbar& scrollbar)
```
I wonder if it's time to retire the term "painter" and just say "scrollerImp".

nmoucht replied:
Changed to scrollerImp

smfr commented on Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
```diff
@@ -954,11 +955,10 @@ - (void)invalidate
         NSScrollerImp *oldVerticalPainter = [m_scrollerImpPair verticalScrollerImp];
-        auto newVerticalPainter = retainPtr([NSScrollerImp scrollerImpWithStyle:newStyle controlSize:(NSControlSize)verticalScrollbar->widthStyle() horizontal:NO replacingScrollerImp:oldVerticalPainter]);
+        auto* verticalScrollbarMac = dynamicDowncast<ScrollbarMac>(verticalScrollbar);
+        verticalScrollbarMac->createScrollerImp(WTFMove(oldVerticalPainter));
```
Do we know it's not null? What if the scrollbar is a custom scrollbar?

nmoucht replied:
There is a check for if its a custom scrollbar above this code.

nmoucht committed:
Refactor Scrollbar to hold NSScrollerImp
https://bugs.webkit.org/show_bug.cgi?id=264014
rdar://117770817

Reviewed by Simon Fraser.

Decouple creation of NSScrollerImps from the global map in
ScrollbarThemeMac. Instead, create ScrollbarMac to hold a reference to the
NSScrollerImp, and change the map to a set of scrollbars (which is still
necessary due to the WebScrollbarPrefsObserver stuff).

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::createNativeScrollbar):
* Source/WebCore/platform/mac/ScrollbarMac.h: Added.
(WebCore::ScrollbarMac::~ScrollbarMac):
* Source/WebCore/platform/mac/ScrollbarMac.mm: Added.
(WebCore::ScrollbarMac::ScrollbarMac):
(WebCore::ScrollbarMac::painterForScrollbar):
(WebCore::ScrollbarMac::createNSScrollerImp):
(WebCore::macScrollbarTheme):
(WebCore::ScrollbarMac::updateNSScrollerImpState):
(WebCore::ScrollbarMac::setNewPainterForScrollbar):
* Source/WebCore/platform/mac/ScrollbarThemeMac.h:
* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::scrollbarMap):
(+[WebScrollbarPrefsObserver appearancePrefsChanged:]):
(WebCore::ScrollbarThemeMac::registerScrollbar):
(WebCore::ScrollbarThemeMac::unregisterScrollbar):
(WebCore::ScrollbarThemeMac::painterForScrollbar):
(WebCore::ScrollbarThemeMac::hasThumb):
(WebCore::ScrollbarThemeMac::minimumThumbLength):
(WebCore::ScrollbarThemeMac::updateEnabledState):
(WebCore::paintScrollbar):
(WebCore::ScrollbarThemeMac::setNewPainterForScrollbar): Deleted.
* Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
(WebCore::ScrollbarsControllerMac::updateScrollerStyle):

Canonical link: https://commits.webkit.org/272044@main
